sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,633 @@
+accordance with the table below.
+
+_Returns_: Information about the kernel that applies when the kernel is
+submitted with the configuration described by the parameter pack `+T...+`.
+
+This extension adds several new queries to this interface, many of which
+already have equivalents in the `kernel_device_specific` or `device`
+namespaces.
+
+NOTE: Redefining these queries is necessary to define `Param::argument_types`.
+If this extension is proposed for a future SYCL standard, such duplication
+could be avoided.
+
+[source, c++]
+----
+namespace ext::oneapi::experimental::info::kernel {
+
+template <uint32_t Dimensions>
+struct max_work_item_sizes;
+
+struct max_work_group_size;
+struct max_num_work_groups;
+
+struct max_sub_group_size;
+struct max_num_sub_groups;
+
+}
+----
+
+[%header,cols="1,5,5,5"]
+|===
+|Kernel Descriptor
+|Argument Types
+|Return Type
+|Description
+
+|`template <uint32_t Dimensions>
+ max_work_item_sizes`
+|`sycl::queue`
+|`id<Dimensions>`
+|Returns the maximum number of work-items that are permitted in each dimension
+ of a work-group, when the kernel is submitted to the specified queue,
+ accounting for any kernel properties or features. If the kernel can be
+ submitted to the specified queue without an error, the minimum value returned
+ by this query is 1, otherwise it is 0.
+
+|`max_work_group_size`
+|`sycl::queue`
+|`size_t`
+|Returns the maximum number of work-items that are permitted in a work-group,
+when the kernel is submitted to the specified queue, accounting for any
+kernel properties or features. If the kernel can be submitted to the specified
+queue without an error, the minimum value returned by this query is 1,
+otherwise it is 0.
+
+|`max_num_work_groups`
+|`sycl::queue`, `sycl::range`, `size_t`
+|`size_t`
+|Returns the maximum number of work-groups, when the kernel is submitted to the
+specified queue with the specified work-group size and the specified amount of
+dynamic work-group local memory (in bytes), accounting for any kernel
+properties or features. If the kernel can be submitted to the specified queue
+without an error, the minimum value returned by this query is 1, otherwise it
+is 0.
+
+|`max_sub_group_size`
+|`sycl::queue`, `sycl::range`
+|`uint32_t`
+|Returns the maximum sub-group size, when the kernel is submitted to the
+specified queue with the specified work-group size, accounting for any kernel
+properties or features. If the kernel can be submitted to the specified queue
+without an error, the minimum value returned by this query is 1, otherwise it
+is 0.
+
+|`max_num_sub_groups`
+|`sycl::queue`, `sycl::range`
+|`uint32_t`
+|Returns the maximum number of sub-groups per work-group, when the kernel is
+submitted to the specified queue with the specified work-group size, accounting
+for any kernel properties or features. If the kernel can be submitted to the
+specified queue without an error, the minimum value returned by this query is
+1, otherwise it is 0.
```

Comment:
> It took me a while to remember this, because it isn't obvious. Do you think it should be documented as a non-normative note?

I think I'd make two changes.  First, I'd separate the queries into two tables.  The first table contains the descriptors that provide limits on how the user must launch the kernel.  The introductory text in this section already explains that the descriptors provide limits that the user must abide by.  The second table contains the descriptors that provide information about the way the implementation launches the kernel (i.e. `max_sub_group_size` and `max_num_sub_groups`).

Second, I'd add a sentence to `max_sub_group_size` noting that this descriptor returns the same value as `get_max_local_range`.  I assume it is important for the user to know this, right?  For example, the host code might allocate some buffer with one element for each sub-group.

Reply:
I've made both of these changes in 14abbedd63ce679819e29f64aaf52b3ccb46a65e.

Structuring things this way made me realize that the name `max_num_sub_groups` didn't make much sense.  By the time the user queries it, the number of sub-groups is fixed.  I think `num_sub_groups` is clearer, so I've made that change in 269386579151e0bf59e5d97fec2f4c932fa65fe0.  `max_sub_group_size` still has a `max` in it because of the connection to `max_local_range`.